src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -0,0 +1,450 @@
+        """
+        self.mgr = dataManager
+        self.logger = logger
+        self.rucioAccount = rucioAccount
+        self.rucioClient = rucioClient
+        self.report = MSPileupReport()
+
+    def monitoringTask(self):
+        """
+        Execute Monitoring task according to the following logic:
+
+        1. Read pileup document from MongoDB with filter active=true
+        2. For each rule id in ruleIds:
+           - query Rucio for that rule id and fetch its state (e.g.: afd122143kjmdskj)
+           - if state=OK, log that the rule has been satisfied and add that RSE to the currentRSEs (unique)
+           - otherwise, calculate the rule completion based on the 3 locks_* field
+        3. now that all the known rules have been inspected, persist the up-to-date pileup doc in MongoDB
+        """
+        spec = {'active': True}
+        docs = self.mgr.getPileup(spec)
+        taskSpec = self.getTaskSpec()
+        asyncio.run(performTasks('monitoring', docs, taskSpec))
+
+    def inactiveTask(self):
+        """
+        Inactive pileup task:
+
+        This task is supposed to look at pileup documents that have been set to
+        inactive. The main goal here is to ensure that there are no Rucio rules
+        left in the system (of course, for the relevant DID and the Rucio
+        account adopted by our microservice). Pileup documents that are updated
+        as a result of this logic should have their data persisted back in
+        MongoDB. A short algorithm for it can be described as follows:
+
+        1. Read pileup document from MongoDB with filter active=false
+        2. for each DID and Rucio account, get a list of all the existent rules
+           - make a Rucio call to delete that rule id, then:
+             - remove the rule id from ruleIds (if any) and remove the RSE name from currentRSEs (if any)
+        3. make a log record if the DID + Rucio account tuple does not have any existent rules
+           - and set ruleIds and currentRSEs to an empty list
+        4. once all the relevant rules have been removed, persist an up-to-date version of the pileup data structure in MongoDB
+        """
+        spec = {'active': False}
+        docs = self.mgr.getPileup(spec)
+        taskSpec = self.getTaskSpec()
+        asyncio.run(performTasks('inactive', docs, taskSpec))
+
+    def activeTask(self, marginSpace=1024**4):
+        """
+        Active pileup task:
+        :param marginSpace: minimum margin space size in bytes to have at RSE to place a dataset
+
+        This task is supposed to look at pileup documents active in the system.
+        Its main goal is to ensure that the pileup DID has all the requested
+        rules (and nothing beyond them), according to the pileup object
+        configuration. Pileup documents that are updated as a result of this
+        logic should have their data persisted back in MongoDB.
+
+		1. Read pileup document from MongoDB with filter active=true
```

Comment:
explanation too much indented.

Reply:
I broke lines to make it fit to the screen, apart from that I do not know how to make it better.